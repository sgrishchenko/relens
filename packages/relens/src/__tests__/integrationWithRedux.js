import { createStore } from 'redux';
import { relensReducer, select, update } from '../index';

describe('integration With Redux', () => {
  test('standard usage flow', () => {
    const EXAMPLE_ACTION_TYPE = 'EXAMPLE_ACTION_TYPE';
    const exampleAction = {
      type: EXAMPLE_ACTION_TYPE,
      payload: 'some payload',
    };

    const initialState = { prop: 'some initial value' };
    const resultState = { prop: 'some state value' };

    const rootReducer = (state = initialState, action) => {
      switch (action.type) {
        case EXAMPLE_ACTION_TYPE: {
          return resultState;
        }
        default: {
          return state;
        }
      }
    };

    const store = createStore(relensReducer(rootReducer));

    expect(store.getState()).toMatchSnapshot('[0] initial state');

    store.dispatch(exampleAction);
    expect(store.getState()).toEqual(resultState);
    expect(store.getState()).toMatchSnapshot('[1] state after the work of the rootReducer');

    const value = 'value';
    const exampleId = 42;
    const relensAction = update(({ id }) => ['some', 'deep', 'prop', id])(value, { id: exampleId });
    store.dispatch(relensAction);
    expect(store.getState().some.deep.prop[exampleId]).toEqual(value);
    expect(store.getState()).toMatchSnapshot('[2] state after the work of the relensReducer');

    const actualStateValue = select(({ id }) => ['some', 'deep', 'prop', id])(store.getState(), { id: exampleId });
    expect(actualStateValue).toBe(value);
  });
});
