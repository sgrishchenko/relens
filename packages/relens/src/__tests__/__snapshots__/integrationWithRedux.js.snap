// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration With Redux standard usage flow: [0] initial state 1`] = `
Object {
  "prop": "some initial value",
}
`;

exports[`integration With Redux standard usage flow: [1] state after the work of the rootReducer 1`] = `
Object {
  "prop": "some state value",
}
`;

exports[`integration With Redux standard usage flow: [2] state after the work of the relensReducer 1`] = `
Object {
  "prop": "some state value",
  "some": Object {
    "deep": Object {
      "prop": Object {
        "42": "value",
      },
    },
  },
}
`;
